File: sdks/python/costobs/metadata.py
"""Metadata resolution with well-defined precedence.

Precedence (later wins): client-level config < request-context < call-site.
None values are dropped at every layer so they never clobber a real value.
"""

from __future__ import annotations

from typing import Any, Dict

from . import context

# Recognized top-level metadata fields that map onto Event columns.
_KNOWN_FIELDS = {
    "environment",
    "team",
    "service",
    "customer_id",
    "user_id",
    "trace_id",
    "feature",
    "prompt_key",
    "prompt_version",
}


def _clean(d: Dict[str, Any]) -> Dict[str, Any]:
    return {k: v for k, v in d.items() if v is not None}


class MetadataResolver:
    """Holds client-level metadata and merges in higher-precedence layers."""

    def __init__(self, client_meta: Dict[str, Any]):
        # Split into known fields vs free-form tags.
        self.client_fields: Dict[str, Any] = {}
        self.client_tags: Dict[str, str] = {}
        for k, v in _clean(client_meta).items():
            if k == "tags" and isinstance(v, dict):
                self.client_tags.update({tk: str(tv) for tk, tv in v.items() if tv is not None})
            elif k in _KNOWN_FIELDS:
                self.client_fields[k] = v
            else:
                self.client_tags[k] = str(v)

    def merge(self, call_meta: Dict[str, Any]) -> Dict[str, Any]:
        """Return merged metadata: client < request-context < call-site."""
        fields: Dict[str, Any] = dict(self.client_fields)
        tags: Dict[str, str] = dict(self.client_tags)

        # Request-context layer.
        ctx = _clean(context.current())
        for k, v in ctx.items():
            if k == "tags" and isinstance(v, dict):
                tags.update({tk: str(tv) for tk, tv in v.items() if tv is not None})
            elif k in _KNOWN_FIELDS:
                fields[k] = v
            else:
                tags[k] = str(v)

        # Call-site layer (highest precedence).
        call = _clean(call_meta)
        call_tags = call.pop("costobs_tags", None)
        for k, v in call.items():
            if k in _KNOWN_FIELDS:
                fields[k] = v
            else:
                tags[k] = str(v)
        if isinstance(call_tags, dict):
            tags.update({tk: str(tv) for tk, tv in call_tags.items() if tv is not None})

        result: Dict[str, Any] = {k: v for k, v in fields.items()}
        result["tags"] = tags
        return result

File: sdks/typescript/src/metadata.ts
/**
 * Metadata resolution with well-defined precedence.
 *
 * Precedence (later wins): client-level config < request-context
 * (AsyncLocalStorage) < call-site. null/undefined values are dropped at every
 * layer so they never clobber a real value.
 */

import { currentContext } from "./context.js";

/** Recognized event fields. Keyed by the canonical (snake_case) event key. */
const KNOWN_FIELDS = new Set([
  "environment",
  "team",
  "service",
  "customer_id",
  "user_id",
  "trace_id",
  "feature",
  "prompt_key",
  "prompt_version",
]);

/**
 * camelCase / call-site aliases mapped to canonical event field names. This
 * lets callers use idiomatic JS (`customerId`, `promptVersion`) while the
 * wire contract stays snake_case.
 */
const FIELD_ALIASES: Record<string, string> = {
  customerId: "customer_id",
  userId: "user_id",
  traceId: "trace_id",
  promptKey: "prompt_key",
  promptVersion: "prompt_version",
};

function canonicalKey(key: string): string {
  return FIELD_ALIASES[key] ?? key;
}

export interface ResolvedMetadata {
  environment?: string;
  team?: string;
  service?: string;
  customer_id?: string;
  user_id?: string;
  trace_id?: string;
  feature?: string;
  prompt_key?: string;
  prompt_version?: string;
  tags: Record<string, string>;
}

/** Call-site metadata as read from the stripped `costobs` field. */
export interface CallSiteMeta {
  feature?: string;
  promptKey?: string;
  promptVersion?: string;
  /** Also accepts snake_case for parity. */
  prompt_key?: string;
  prompt_version?: string;
  tags?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface ClientMeta {
  environment?: string;
  team?: string;
  service?: string;
  tags?: Record<string, unknown>;
  [key: string]: unknown;
}

function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/** Holds client-level metadata and merges in higher-precedence layers. */
export class MetadataResolver {
  private readonly clientFields: Record<string, string> = {};
  private readonly clientTags: Record<string, string> = {};

  constructor(clientMeta: ClientMeta = {}) {
    for (const [rawKey, value] of Object.entries(clientMeta)) {
      if (value === undefined || value === null) continue;
      const key = canonicalKey(rawKey);
      if (key === "tags" && isPlainObject(value)) {
        this.mergeTags(this.clientTags, value);
      } else if (KNOWN_FIELDS.has(key)) {
        this.clientFields[key] = String(value);
      } else {
        this.clientTags[key] = String(value);
      }
    }
  }

  private mergeTags(
    into: Record<string, string>,
    from: Record<string, unknown>,
  ): void {
    for (const [k, v] of Object.entries(from)) {
      if (v !== undefined && v !== null) into[k] = String(v);
    }
  }

  /** Return merged metadata: client < request-context < call-site. */
  merge(callMeta: CallSiteMeta = {}): ResolvedMetadata {
    const fields: Record<string, string> = { ...this.clientFields };
    const tags: Record<string, string> = { ...this.clientTags };

    // Request-context layer (AsyncLocalStorage).
    for (const [rawKey, value] of Object.entries(currentContext())) {
      if (value === undefined || value === null) continue;
      const key = canonicalKey(rawKey);
      if (key === "tags" && isPlainObject(value)) {
        this.mergeTags(tags, value);
      } else if (KNOWN_FIELDS.has(key)) {
        fields[key] = String(value);
      } else {
        tags[key] = String(value);
      }
    }

    // Call-site layer (highest precedence).
    for (const [rawKey, value] of Object.entries(callMeta)) {
      if (value === undefined || value === null) continue;
      const key = canonicalKey(rawKey);
      if (key === "tags" && isPlainObject(value)) {
        this.mergeTags(tags, value);
      } else if (KNOWN_FIELDS.has(key)) {
        fields[key] = String(value);
      } else {
        tags[key] = String(value);
      }
    }

    return { ...fields, tags } as ResolvedMetadata;
  }
}
